# 第２回TAからの練習問題  （解答例）
テーマ１. **数学力とプログラミング力をうまく組み合わせて、知りたい値を計算によって求める。**  
テーマ２. **Pythonの計算速度の目安を知る。**

### 問題１  
数列 $a_{n} = 2^{n}$について考える。  $a_{1},a_{2},a_{3},\ldots,a_{2022}$の2022個の数のうち、最高位の数字が４であるものの個数が知りたい。  
$a_{1},a_{2},a_{3},\ldots,a_{k}$までの$k$個の数のうち、最高位の数字が4であるものの個数を返す関数number_of_top4($k$)を作り、number_of_top4(2022)を計算してみよう。  
方法はたくさん考えられるので、そのうち1つ以上の方法でプログラムを書いてみよう。  
以下の(参考)を参考にしてもよい。

In [1]:
#解答例1
#実際に計算→型を文字列に変えて１文字目の文字を確認する。
def number_of_top4_v1(k):
    count = 0
    for i in range(1,k+1):
        x = 2**i
        x = str(x)
        if(x[0]=='4'):
            count += 1
    return count

In [2]:
number_of_top4_v1(2022)

197

In [3]:
#解答例2
#対数になおして調べる
import math
def number_of_top4_v2(k):
    count=0
    for i in range(1,k+1):
        x = i * math.log10(2)
        x = x - int(x)
        if(math.log10(4)-10**(-13) < x < math.log10(5)-10**(-13)):
            count += 1
    return count

In [4]:
number_of_top4_v2(2022)

197

In [5]:
#解答例3
#実際に計算→1桁の数になるまで10で割っていく。
def number_of_top4_v3(k):
    count = 0
    for i in range(1,k+1):
        x=1
        for _ in range(i):
            x = x * 2
            while(x>10):
                x=x / 10
        x = int(x)
        if(x==4):
            count +=1
    return count

In [6]:
number_of_top4_v3(2022)

197

### 問題２（応用）
$a_{1},a_{2},a_{3},\ldots,a_{10^{12}}(=a_{1000000000000})$までの$10^{12}$（1兆）個の数のうち、最高位の数字が4であるものの個数を求めよう。  兆
ただし、計算に用いた値の有効数字は気にしなくてよい。(float型の有効数字は15桁ほどで12～13桁なら正確に表示できている。)  
さて、先ほど作ったnumber_of_top4(10**12)で求められるだろうか？

数学力や発想力などを駆使して、問題2の解答が求められるように関数number_of_top4を改良しよう。  
$a_{1},a_{2},a_{3},\ldots,a_{10^{12}}(=a_{1000000000000})$までの$10^{12}$個の数のうち、最高位の数字が4であるものの個数を求めよう。  

In [11]:
def top(k):
    if (k==1):
        return 2
    elif (k==2):
        return 4
    elif (k==3):
        return 8
    elif (k*math.log10(2) - int(k*math.log10(2)) < math.log10(2) ):
        return 1
    elif (k*math.log10(2) - int(k*math.log10(2)) < math.log10(4) ):
         return 2
    elif (k*math.log10(2) - int(k*math.log10(2)) < math.log10(5) ):
         return 4
    elif (k*math.log10(2) - int(k*math.log10(2)) < math.log10(8) ):
         return 5
    else:
         return 8

In [12]:
def number_of_top4_fast(k):
    a=0
    if(top(k)==2):
        a=1
    elif(top(k)==4):
        a=2
    elif(top(k)==5):
        a=2
    elif(top(k)==8):
        a=3
    
    answer = k - a - 3*int(k*math.log10(2))
    if( (top(k)==4) | (top(k)==8) ):
        answer +=1
    return answer

In [13]:
print(number_of_top4_fast(10**12))

96910013009


問題1の答え　197個  
問題2の答え　96910013009個（969億1001万3009個）

念のため、number_of_top4_v2(k)とnumber_of_top4_fast(k)の出力に違いがないことを確認しておこう。

In [15]:
for i in range(10**4):
    if(number_of_top4_v2(i) != number_of_top4_fast(i)):
        print(i)
print("実行終了")

実行終了


以下、number_of_top4_fastの計算の方法の説明を簡単に行う。  
ここでは、$a_{1},a_{2},a_{3},\ldots,a_{2022}$の2022個の数のうち、最高位の数字が4であるものの個数を求める。
これを一般化することで、number_of_top4_fastになる。

まず、$2^{2022}$の最高位の数と桁数を求める。$\log_{10}2^{2022}=2022×\log_{10}2=608.68\cdots$であり、$\log_{10}4=0.602,\log_{10}5=0.699$であることをふまえると、桁数は609で最高位の数は4であるとわかる。下の図を参考にすると、$2^{2020}$は桁数が609で最高位の数は1であるとわかる。
最高位の数が4を経由して1桁上がる回数をa,最高位の数が4を経由せずに1桁上がる回数をbとおく。
桁数を考えると、$a_{2020}$までで608回桁が上がるので$a+b=608\cdots…$①が成立する。
2をかけた回数を考えると$4a+3b=2020\cdots$②が成立する。
①、②より連立方程式を解いて$a=196,b=412$となる。よって、$a_{2020}$までで最高位が4の数は196個、それに$a_{2022}$の分を足して求める個数は197個とわかる。$a_{1兆}$の時も同様に計算すればよい。

![最高位の遷移の様子](A2figure.png)

In [8]:
2022*math.log10(2)

608.68265123257